src/ica_api.c

```diff
@@ -2761,6 +2767,9 @@ unsigned int ica_aes_ctr(const unsigned char *in_data, unsigned char *out_data,
 #ifdef ICA_FIPS
 	if (fips >> 1)
 		return EACCES;
+	if (ctr_width < (AES_BLOCK_SIZE*8) &&
+		data_length / AES_BLOCK_SIZE > (1ULL << ctr_width))
```

Comment:
This does not work!  `AES_BLOCK_SIZE` is 16.  So we check that we shift less than 16*8 bits, but we only have a 64bit value.  So every shift >= 64 does not make sense.  Since `data_length / AES_BLOCK_SIZE` is strictly less than 2^64, why don't you replace the check with a hard comparison against 64 and then the shift?  Also, what about padding?  If `data_length` is not a multiple of `AES_BLOCK_SIZE`, please use this check:
```
if (ctr_width < 64U && ((data_length / AES_BLOCK_SIZE) + !!(data_length % AES_BLOCK_SIZE)) > (1UL << ctr_width))
```
This makes sure that `ctr_width` greater equal 64 bits is not a problem (since we cannot have that many blocks), and that the number of blocks needed (including possibly padded blocks via the modulo and double negation) still does not cause repeated counter values.

Similar check could be used for DES above.

Reply:
Yes this makes sense:
```
if (ctr_width < (64*8) &&
		data_length / 64 > (1ULL << ctr_width))
```
In fact, using the DES_BLOCK_SIZE constant in the ica_3des_ctr routine (btw I removed the check completely for ica_des_ctr, because DES is blocked in fips mode anyway) is not correct. By coincidence, DES_BLOCK_SIZE is 64 bits, but it means in fact sizeof(unsigned long).

Reply:
I don't understand why you need `sizeof(unsigned long)` here?  Isn't the FIPS requirement for CTR-mode that the counter is not allowed to have the same value for different blocks?  So we need to divide the (possibly padded) data length by the blocksize and check if there are enough counter bits to represent that number.  That was what I tried to represent in the formula above.  

Of course, this only works if all data is passed in one go to this function.  If the function can be called multiple times in a streaming fashion, then we still need another additional check.  I am not sure about the usage pattern of this function, but I assume it is only used to de- or encrypt a whole message at once.